### ISSUE TITLE
GeoshapePoint multiples issues

### ISSUE THREAD

**Issue by ganjatealq** [elastic/elasticsearch-net#1908]
We are having 2 issues with querying GeoshapePoint with ES 1.7.1, and Nest 1.7.0

Here our C# : 

```
var result = client.Search<Polygon>(
                s => s.Index(IndexNames.Ref)
                    .Query(q => q.Filtered(
                        f => f.Filter(g =>
                            g.GeoShapePoint("VertexValues",
                                d => d.Relation(GeoShapeRelation.Within).Coordinates(coordinates)
                                      )
                                )
                            )
                        ));
```

Here is the json generated : 

```
{
  "query": {
    "filtered": {
      "filter": {
        "geo_shape": {
          "VertexValues": {
            "shape": {
              "coordinates": [
                48.864686,
                2.351616
              ],
              "Type": "point"
            },
            "relation": "within"
          }
        }
      }
    }
  }
}
```

Problems are : 
- "**T**ype" should be in lower case
- "relation": "within" should be inside the "shape" brackets

**Comment by ganjatealq** [comment]
Any ETA on Next Legacy Release ?

**Comment by Mpdreamz** [comment]
Hi @ganjatealq we discussed this today and we will tackle a fix and patch release this week. 

Thanks for the bump :+1:

**Comment by russcam** [comment]
Hey @ganjatealq, I can't reproduce this issue; I've tried with NEST versions 1.7.0 down to 1.6.0 but always get the expected output

```javascript
{
  "query": {
    "filtered": {
      "filter": {
        "geo_shape": {
          "VertexValues": {
            "shape": {
              "coordinates": [
                48.864686,
                2.351616
              ],
              "type": "point"
            },
            "relation": "within"
          }
        }
      }
    }
  }
}
```

Some questions to see what might be causing this

1. How did you get the json for the query? 
2. Do you get an error back from Elasticsearch; if so, what is it?
3. Do you use Json.Net for serialization; if so, which version, and do you make any modifications to the default?

If you'd be able to supply a complete example to replicate this (attach as a zip), I can look further.

`relation` should be a separate property to `shape` on the object assigned to the field name. Take a look at the [`GeoShapeFilterParser` from Elasticsearch 1.7.1](https://github.com/elastic/elasticsearch/blob/v1.7.1/src/main/java/org/elasticsearch/index/query/GeoShapeFilterParser.java#L52)

**Comment by ganjatealq** [comment]
hi @russcam 

1) I used Fiddler to get it.
2) I got one yes, see capture ![image](https://cloud.githubusercontent.com/assets/17412391/15703124/0e7a8a4e-27e4-11e6-88be-307eb4c5245e.png)
3) We use Newtonsoft.Json version 7.0.0.0, and no modification

Unfortunately, I cannot supply a complete example..

 I'm not sure to understand your last sentence.
To bypass this bug, we manage to use the following code : 

            var jsonString = "{\"filtered\":{\"filter\":{\"geo_shape\":{\"VertexValues\":{\"shape\":{\"coordinates\":[" + x.ToString(CultureInfo.InvariantCulture) + "," + y.ToString(CultureInfo.InvariantCulture) + "],\"type\":\"point\",\"relation\":\"within\"}}}}}}";

            var result = client.Search<Polygon>(
                s => s.Index(IndexNames.Ref)
                    .QueryRaw(jsonString));

And it works (while being ugly code). 
Using this json string, copy/paste in ES it works perfectly : 
![image](https://cloud.githubusercontent.com/assets/17412391/15703578/5864b02e-27e6-11e6-9f90-43b6a94d8ee1.png)

**Comment by russcam** [comment]
Ok, I think I know what is going on here. @ganjatealq, would you be able to share your connection settings? Are you using `.SetDefaultPropertyNameInferrer` and setting it to *not* lowercase properties?

**Comment by russcam** [comment]
The issue here is that `GeoShape` does not implement `IGeoShape` - the interface has a `JsonPropertyAttribute` on `Type` property to use the property name `"type"` when serializing. All geo shapes inherit from `GeoShape` and implement an interface that derives from `IGeoShape`, but the contract resolver is not configured to pick up the `JsonPropertyAttribute` on the `IGeoShape` implementation so will roll with normal Json.Net serialization behaviour for the inherited `Type` property from `GeoShape`.

This **is** a problem but *it doesn't normally manifest itself* in queries because the default property name inferrer for NEST will camel case all property names. If the inference for property names is changed on Connection Settings though (which I suspect it might be in your case) like so

```c#
var settings = new ConnectionSettings()
    // this will cause "Type" to be serialized for geo queries and filters :(
    .SetDefaultPropertyNameInferrer(p => p)
```

Then the `Type` property ends up being serialized as `"Type"` :cry: This is the exception you have in Fiddler.

With regards to `"relation"`, it is serialized in the correct place, but I think your query is not doing what you anticipate; the filter is asking for all shapes in the index that are within the given point coordinates in the query input. Based on the polygon result that I can see in the image, this would not be _within_ the point and so you would get no results (ignoring the `Type` error for the moment). 

Where you have `"relation"` in `jsonString` above is *ignored* when deserializing the request in Elasticsearch, and the filter [will use the default relation which is `INTERSECT`](https://github.com/elastic/elasticsearch/blob/v1.7.1/src/main/java/org/elasticsearch/index/query/GeoShapeFilterParser.java#L83), which will return shapes in the index that intersect with the provided point. You will see the same results if  you use

```c#
	var response = client.Search<Polygon>(s => s
		.QueryRaw(@"{
			""filtered"": {
				""filter"": {
					""geo_shape"": {
						""VertexValues"": {
							""shape"": {
								""coordinates"": [
				                  48.864686,
				                  2.351616
			          			],
			          		""type"": ""point""
			        		},
							""relation"": ""intersects""
			      		}
			    	}
			  	}
			}
		}")
	);
``` 

So I think you want `INTERSECT` in this case, based on expectation. Conversely, if you had a bunch of points in the index and wanted to return results that are within a given polygon, the `WITHIN` relation would be what you want.

I've fixed the `Type` property issue and will push out a new release today. Thanks so much for reporting and providing additional details :+1:

**Comment by russcam** [comment]
This is now [fixed in NEST 1.8.2](https://www.nuget.org/packages/NEST/1.8.2) :+1:

**Comment by ganjatealq** [comment]
hi @russcam 

Thanks for fixing the Type issue :)

As for the "relation: within", it seems that we misunderstood its using in our case.
We wanted to know which polygons in our index contain our point coordinates (in query), and querying the other way around ;)
We removed it and it works.

thanks again
